src/types.md

```diff
@@ -236,3 +236,311 @@ fn main() {
+}
+```
+
+(the type system would prevent these operations happening in parallel).
+
+*Marker traits*. Indicate that a type meets certain invariants, so subsequent
+users of that type don't need to check at runtime. A common example is to
+indicate that a type is safe to serialize into some bytestream.
+
+*Enums as state machines*. Each enum variant is a state and stores data
+associated with that state. There simply is no possibility that the data can
+get out of sync with the state.
+
+```rust
+enum ElectionState {
+  RaisingDonations { amount_raised: u32 },
+  DoingTVInterviews { interviews_done: u16 },
+  Voting { votes_for_me: u64, votes_for_opponent: u64 },
+  Elected,
+  NotElected,
+};
+```
+
+(A more heavyweight approach here is to define types for each state, and
+allow valid state transitions by taking the previous state by-value and
+returning the next state by-value.)
+
+```rust
+struct Seed { water_available: u32 }
+struct Growing { water_available: u32, sun_available: u32 }
+struct Flowering;
+struct Dead;
+
+enum PlantState {
+  Seed(Seed),
+  Growing(Growing),
+  Flowering(Flowering),
+  Dead(Dead)
+}
+
+impl Seed {
+  fn advance(self) -> PlantState {
+    if self.water_available > 3 {
+      PlantState::Growing(Growing { water_available: self.water_available, sun_available: 0 })
+    } else {
+      PlantState::Dead(Dead)
+    }
+  }
+}
+
+impl Growing {
+  fn advance(self) -> PlantState {
+    if self.water_available > 3 && self.sun_available > 3 {
+      PlantState::Flowering(Flowering)
+    } else {
+      PlantState::Dead(Dead)
+    }
+  }
+}
+
+impl Flowering {
+  fn advance(self) -> PlantState {
+    PlantState::Dead(Dead)
+  }
+}
+
+impl Dead {
+  fn advance(self) -> PlantState {
+    PlantState::Dead(Dead)
+  }
+}
+
+impl PlantState {
+  fn advance(self) -> Self {
+    match self {
+      Self::Seed(seed) => seed.advance(),
+      Self::Growing(growing) => growing.advance(),
+      Self::Flowering(flowering) => flowering.advance(),
+      Self::Dead(dead) => dead.advance(),
+    }
+  }
+}
+
+// we should probably find a way to inject some sun and water into this
+// state machine or things are not looking rosy
+```
+
+## When should I use `Rc` versus `Arc`?
+
+Never expose `Rc` in APIs. `Arc` is the right choice nearly always. (You
+might want to look at the [archery crate](https://docs.rs/archery/latest/archery/)).
+
+## What should I do instead of inheritance?
+
+Use composition. Sometimes this results in more boilerplate, but it avoids
+a raft of complexity.
+
+Specifically, for example:
+* you might include the "superclass" struct as a member of the subclass
+  struct;
+* you might use an enum with different variants for the different possible
+  "subclasses".
+
+Usually the answer is obvious: it's unlikely that your Rust code is structured
+in such a way that inheritance would be a good fit anyway.
+
+> I've only missed inheritance when actually implementing languages which
+> themselves have inheritance - MG.
```

Comment:
But that's exactly the situation readers of this FAQ find themselves in, so, maybe not helpful.

Reply:
Italicized _implementing_ for additional clarity